### PR TITLE
Change viewportTopOffset config. Make viewportTopOffset writable in runtime

### LINKED ISSUE #9672: Allow modifying toolbar's viewportTopOffset in runtime
## 📝 Provide a description of the improvement

Currently, `config.toolbar.viewportTopOffset` can only be set when creating the editor.

The value then ends up in `editor.ui.view.stickyPanel.viewportTopOffset` that's readonly.

Could we make this property writable? Or, maybe even better, add a method/getter in the `editor.ui.view` or somewhere deeper? 

The problem I see is that the (in the runtime) `viewportTopOffset` is not a configuration of `editor.ui.view.toolbar` but `editor.ui.view.stickyPanel` which is rather poorly named (seems to have nothing to do with the toolbar). Even if we'll make it writable, discoverability will be relatively poor.

That might be an argument for adding a method to control this property somewhere up the chain (e.g. in `editor.ui.view`) but I dont' want to overcomplicate this so I have mixed feelings.

cc @oleq 

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by dawidurbanski**
WIP (check additional information)

### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Type: Change viewportTopOffset config. Make viewportTopOffset writable in runtime. Closes #9672.

---

### Additional information

I added couple `TODO` comments to point out some stuff that could be improved in this PR IMO.

@oleq if you would find some time to check this one it would be awesome.

These two manual tests are best to see this in action:

```
yarn manual --files=editor-classic/manual/tickets/9672/1.js,editor-inline/manual/tickets/9672/1.js
```

Follow ups on this one:

- [x] ~~Update all our tests to use the new config (`ui.viewportOffset.top` instead of `toolbar.viewportTopOffset`)~~ done in d114b6b
- [x] ~~Update all docs snippets to use the new config~~ done in e4429e5
- [x] ~~Update docs to reflect this change~~ done in be2339d
- [x] ~~Update TODOs when discussed~~ done in 39597fd and 83b130c

### CODE REVIEW

**Review by oleq — COMMENTED** [review]
I think this is the right direction 👍

I read your comments @dawidurbanski but instead of replying under individual lines of code, I decided to fork your solution (`ck/9672-modify-viewporTopOffset-in-runtime-v2`) with a [single commit](https://github.com/ckeditor/ckeditor5/commit/6e76b5a1d09f2912c901d225ccc2fbf40633cb79) so it is easier to understand what direction I think would be the best. And besides, unlike GH comments, you can run it straight away 😀

In short, 

1. to avoid code duplication, I'm for moving the [offset normalization to `EditorUI`](https://github.com/ckeditor/ckeditor5/commit/6e76b5a1d09f2912c901d225ccc2fbf40633cb79#diff-8f1cc4aa3b82dd2712c650152a49f78304aa1c20bf325c62f6c2ec0858708654R60). All editors with the UI inherit from this class so it's a safe spot IMO.
   * this makes the runtime config available for developers under `editor.ui.viewportOffset`
   * alternatively: this could be moved from `EditorUI` to `EditorUIView`. The only drawback is that the view needs to obtain the editor config via constructure which is, um..., an antipattern (our architecture assumes UI views are completely decoupled from the editor), but I guess this would also work. The property would be `editor.ui.view.viewportOffset`. 
   * now that I think about it, `EditorUIView` could be a better place because `EditorUI` is a controller (glue) and this whole situation is just pure geometry (presentation). All editors with the UI use `EditorUIView` at some point. WDYT?
2. speaking of `viewportOffset` (instead of `viewportTopOffset`), I think it would be best if we made it future-proof and, as you correctly assumed in your code, used an object with `top`, `bottom`, `left`, `right` properties. For now, only the `top` makes sense but this will change one day.
   * the implementation in the twin issue (the one with the "sticky position") allows restricting the viewport Rect from all sides, anyway, so these changes would work together out-of-the-box.
3. moving on, I bound [`editor.ui.view.stickyPanel.viewportTopOffset` (Classic)](https://github.com/ckeditor/ckeditor5/commit/6e76b5a1d09f2912c901d225ccc2fbf40633cb79#diff-f1cff316ea0a20ab6f81df4c5fc96be3bab302b49e04c170bf7c5d961c839c41R145) and [`InlineEditorUIView.viewportTopOffset` (Inline)](https://github.com/ckeditor/ckeditor5/commit/6e76b5a1d09f2912c901d225ccc2fbf40633cb79#diff-99b6f8366a55057302812d639ce384164db9b0ce7cb52f56ea80959a87975cc8R128) to `EditorUI#viewportOffset`.
   * this way, the whole viewport offset situation is transparent to developers. They don't need to worry about which editor type is used and which property should be changed in the runtime. There's a single source of truth for all editors with UI.

**Comment by dawidurbanski** [comment]
Thanks @oleq !

> now that I think about it, EditorUIView could be a better place because EditorUI is a controller (glue) and this whole situation is just pure geometry (presentation). All editors with the UI use EditorUIView at some point. WDYT?

Let's stick to the the `EditorUI` with this. It felt awkward in the `UIView` for me.

I also added some missing pieces, small refactoring, tests, some initial docs, and proper comments in 83b130c commit.

The only thing left now is to update rest of the docs including snippets and it should be job done. I'm going to complete it soon.

File: docs/features/toolbar.md [hunk 1]
```diff
@@ -73,7 +72,11 @@ toolbar: {
 
  * **`removeItems`** &ndash; An array of toolbar item names. With this setting you can modify the default toolbar configuration without the need of defining the entire list (you can specify a couple of buttons that you want to remove instead of specifying all the buttons you want to keep). If, after removing an item, toolbar will have two or more consecutive separators (`'|'`), the duplicates will be removed automatically.
 
- * **`viewportTopOffset`** &ndash; The offset (in pixels) from the top of the viewport used when positioning a sticky toolbar. Useful when a page with which the editor is being integrated has some other sticky or fixed elements (e.g. the top menu). Thanks to setting the toolbar offset, the toolbar will not be positioned underneath or above the page's UI.
+ * **`viewportTopOffset` (deprecated)** &ndash; The offset (in pixels) from the top of the viewport used when positioning a sticky toolbar. Useful when a page with which the editor is being integrated has some other sticky or fixed elements (e.g. the top menu). Thanks to setting the toolbar offset, the toolbar will not be positioned underneath or above the page's UI.
```
Comment (Reinmar) [inline docs/features/toolbar.md:75]: I'm afraid this bit needs to be rethought. Most of the people who will read this guide ever will not know that this option ever existed. So, it makes sense to only briefly mentioned it that it wasthere or not even mention at all.

Instead, we need to think on how to add a section about `config.ui.viewportSth`. It might be a new section (but it's unclear to me how it could be titled... "Configuring the viewport dimensions"? "Limiting the viewport area"?) or we can just mention this property here like it was part of toolbar configuration. No one said that toolbar config cannot be outside `config.toobar` too. The third option will be to use `<info-box>` to detach this information a bit from its surrounding.

tl;dr: When writing docs (not migration guides) focus on what's available currently.

File: packages/ckeditor5-core/src/editor/editorconfig.jsdoc [hunk 2]
```diff
@@ -167,9 +165,12 @@
  *
  *	Line break will work only in the extended format when `shouldNotGroupWhenFull` option is set to `true`.
  *
- * * **`toolbar.viewportTopOffset`** &ndash; The offset (in pixels) from the top of the viewport used when positioning a sticky toolbar.
+ * * **`toolbar.viewportTopOffset` (deprecated)** &ndash; The offset (in pixels) from the top of the viewport used when positioning a sticky toolbar.
```
Comment (Reinmar) [inline packages/ckeditor5-core/src/editor/editorconfig.jsdoc:168]: This is fine that it's mentioned here. API docs are the place to cover deprecated things.

File: packages/ckeditor5-core/src/editor/editorconfig.jsdoc [hunk 3]
```diff
@@ -297,3 +298,28 @@
  *
  * @member {String} module:core/editor/editorconfig~EditorConfig#placeholder
  */
+
+/**
+ * The editor UI configuration.
```
Comment (Reinmar) [inline packages/ckeditor5-core/src/editor/editorconfig.jsdoc:303]: I miss a link to toolbar configuration guide where this option will be explained in a context of what it affects.

File: packages/ckeditor5-core/src/editor/editorui.js [hunk 4]
```diff
@@ -12,8 +12,7 @@
 import ComponentFactory from '@ckeditor/ckeditor5-ui/src/componentfactory';
 import FocusTracker from '@ckeditor/ckeditor5-utils/src/focustracker';
 
-import EmitterMixin from '@ckeditor/ckeditor5-utils/src/emittermixin';
-import mix from '@ckeditor/ckeditor5-utils/src/mix';
+import { ObservableMixin, mix } from '@ckeditor/ckeditor5-utils';
```
Comment (Reinmar) [inline packages/ckeditor5-core/src/editor/editorui.js:15]: Revert please. It's not consistent with how the rest of the imports look here.

I started wondering though, what's the expected way to write these imports. I need to fwd this to the devops team because this deserves to be linted.

Comment (Reinmar) [reply packages/ckeditor5-core/src/editor/editorui.js:15]: Reported: #10375

File: packages/ckeditor5-core/src/editor/editorui.js [hunk 5]
```diff
@@ -53,6 +52,14 @@ export default class EditorUI {
 		 */
 		this.focusTracker = new FocusTracker();
 
+		/**
+		 * Stores viewport offsets from every direction.
```
Comment (Reinmar) [inline packages/ckeditor5-core/src/editor/editorui.js:56]: I'm afraid that this may be severely unclear :D What is a viewport offset? How it's used? Giving an example could be enough.

File: packages/ckeditor5-core/src/editor/editorui.js [hunk 6]
```diff
@@ -53,6 +52,14 @@ export default class EditorUI {
...
+		 *
+		 * @readonly
```
Comment (Reinmar) [inline packages/ckeditor5-core/src/editor/editorui.js:58]: This will not appear in the docs AFAIK. You need to use `@member` if that's set by `this.set()`. JSDoc is able to recognize `this.x = 1`, but not this.

File: packages/ckeditor5-core/src/editor/editorui.js [hunk 7]
```diff
@@ -172,6 +179,46 @@ export default class EditorUI {
 		return this._editableElementsMap;
 	}
 
+	/**
+	 * Returns viewport offsets object
```
Comment (Reinmar) [inline packages/ckeditor5-core/src/editor/editorui.js:183]: Use backticks to wrap code fragments.

File: packages/ckeditor5-core/src/editor/editorui.js [hunk 8]
```diff
@@ -172,6 +179,46 @@ export default class EditorUI {
...
+	 * { top: Number, right: Number, bottom: Number, left: Number }
+	 * Only top property is currently supported
```
Comment (Reinmar) [inline packages/ckeditor5-core/src/editor/editorui.js:185]: Missing:

*   A period at the end.
*   Blank link to separate paragraphs of text.

File: packages/ckeditor5-core/src/editor/editorui.js (hunk 4 above)
Comment (dawidurbanski) [reply packages/ckeditor5-core/src/editor/editorui.js:15]: @oleq  made this change, not sure thee reason behind. I'll make it the old way :)

File: packages/ckeditor5-core/src/editor/editorui.js (hunk 7 above)
Comment (Reinmar) [reply packages/ckeditor5-core/src/editor/editorui.js:183]: Missing `:` at the end. Also, the code snippet deserves to be a block snippet, so a triple backtick.

tldr something like:

````
Returns viewport offsets object:

```js
{ ... }
```

Only something.
````

File: packages/ckeditor5-core/src/editor/editorui.js [hunk 9]
```diff
@@ -172,6 +179,46 @@ export default class EditorUI {
...
+		const editor = this.editor;
+		const viewportOffsetConfig = editor.config.get( 'ui.viewportOffset' );
+
+		if ( viewportOffsetConfig ) {
+			return viewportOffsetConfig;
+		}
+
+		const legacyOffsetConfig = editor.config.get( 'toolbar.viewportTopOffset' );
+
+		// Fall back to deprecated toolbar config
```
Comment (Reinmar) [inline packages/ckeditor5-core/src/editor/editorui.js:200]: End sentences with periods.

File: packages/ckeditor5-core/src/editor/editorui.js [hunk 10]
```diff
@@ -172,6 +179,46 @@ export default class EditorUI {
...
+		if ( legacyOffsetConfig ) {
```
Comment (Reinmar) [inline packages/ckeditor5-core/src/editor/editorui.js:201]: OK, I have very mixed feelings seeing a warning in a `getFoo()` place (a getter for now, but can be a function too). It's just not part of the "reading things logic" to me. However, this is a deprecation notice and after giving it more thought I'm ok after having it here.

File: packages/ckeditor5-core/tests/editor/editorui.js [hunk 11]
```diff
@@ -202,4 +202,32 @@ describe( 'EditorUI', () => {
 			sinon.assert.calledWithMatch( stub, 'editor-ui-deprecated-editable-elements' );
 		} );
 	} );
+
+	describe( '_viewportOffset()', () => {
```
Comment (Reinmar) [inline packages/ckeditor5-core/tests/editor/editorui.js:206]: If it's private it must not be accessed by other classes. See https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/code-style.html#visibility-levels.

File: packages/ckeditor5-core/src/editor/editorui.js [hunk 12]
```diff
@@ -172,6 +179,46 @@ export default class EditorUI {
...
 
+	/**
+	 * Returns viewport offsets object
+	 * { top: Number, right: Number, bottom: Number, left: Number }
+	 * Only top property is currently supported
+	 *
+	 * @private
+	 * @return {Object}
+	 */
+	get _viewportOffset() {
```
[the remark this answers is not in the document]
Comment (Reinmar) [reply packages/ckeditor5-core/src/editor/editorui.js:190]: OTOH, if you need to expose it for test purposes, which we sometimes do to make them simpler, it can be a method. Just avoid such complex getters. `_readViewportOffsetFromConfig()` would work better for me.

File: packages/ckeditor5-core/tests/editor/editorui.js (hunk 11 above)
Comment (Reinmar) [reply packages/ckeditor5-core/tests/editor/editorui.js:206]: Also, I didn't yet see tests, but do we have an integration test for the warning? I can see the second test here, but what if `_viewportOffset` is never accessed for some reason? Who ensures that it is?

File: packages/ckeditor5-core/src/editor/editorui.js (hunk 6 above)
Comment (Reinmar) [reply packages/ckeditor5-core/src/editor/editorui.js:58]: Also, you need to define the type of this property and please provide some example on how to override it.
As for the type, you can go the easy way and define it as `Object` and explain what keys it has in the text.

**Review by Reinmar — CHANGES_REQUESTED** [review]
OK, I'm done with my review. The change works - the top offset is writable in runtime.

There are two things that I'd want to add:

* The migration guide is necessary and it doesn't need to be part of this ticket, but we need to do it. Please open a followup and ping me about it.
* It's a pity that the toolbar doesn't change its position once you update `editor.ui.viewportOffset`. I also checked if calling `editor.ui.update()` works but it doesn't. Could you make a quick check whether there's an option to update (recalculate) the position of it on demand. If so, please report a followup ticket too.
*

File: packages/ckeditor5-core/src/editor/editorconfig.jsdoc (hunk 3 above)
Comment (dawidurbanski) [reply packages/ckeditor5-core/src/editor/editorconfig.jsdoc:303]: The idea here is to not connect it with the toolbar at all. It may affect anything, not just the toolbar.

Also, the `ui` namespace may be used to add any other `ui` related config stuff in the future that may not have anything to do with the toolbar at all.

Since we are completely removing the `vieportTopOffset` from the toolbar config docs, it may be confusing IMHO.

File: docs/features/toolbar.md (hunk 1 above)
Comment (dawidurbanski) [reply docs/features/toolbar.md]: Removed

File: packages/ckeditor5-core/src/editor/editorconfig.jsdoc (hunk 3 above)
Comment (dawidurbanski) [reply packages/ckeditor5-core/src/editor/editorconfig.jsdoc:303]: I linked `ui.viewportOffset` config (in a note) to both places that use the `top` value and vice-versa. I liked both of these places to the `ui` config.

If you are about to build the docs locally, you'll find these under links:

Config: http://127.0.0.1:8080/ckeditor5/29.1.0/api/module_core_editor_editorconfig-EditorConfig.html#member-ui

Inline Editor: http://127.0.0.1:8080/ckeditor5/29.1.0/api/module_editor-inline_inlineeditoruiview-InlineEditorUIView.html#member-viewportTopOffset

Sticky Panel View: http://127.0.0.1:8080/ckeditor5/29.1.0/api/module_ui_panel_sticky_stickypanelview-StickyPanelView.html#member-viewportTopOffset

File: packages/ckeditor5-core/src/editor/editorui.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (dawidurbanski) [reply packages/ckeditor5-core/src/editor/editorui.js:195]: After some thought I realised what actually this method do is to normalize viewportOffset config. Shall I maybe extract it to a separate (even public) function similar to [normalizeToolbarConfig()](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-ui/src/toolbar/normalizetoolbarconfig.js)?

What do you think @Reinmar ?

[the remark this answers is not in the document]
Comment (Reinmar) [reply packages/ckeditor5-core/src/editor/editorui.js:195]: It makes sense to extract it if it's going to be reused somewhere else. If not, keep it local cause KISS.

File: packages/ckeditor5-core/tests/editor/editorui.js (hunk 11 above)
Comment (dawidurbanski) [reply packages/ckeditor5-core/tests/editor/editorui.js]: I changed it to private method, and updated tests to test against `ui.viewportOffset` instead of this private method.

> Also, I didn't yet see tests, but do we have an integration test for the warning?

Yes: [here](https://github.com/ckeditor/ckeditor5/blob/b609fcfa4502eeb3bfd28aec115f5f6c0121d0a6/packages/ckeditor5-editor-classic/tests/classiceditorui.js#L89) and [here](https://github.com/ckeditor/ckeditor5/blob/b609fcfa4502eeb3bfd28aec115f5f6c0121d0a6/packages/ckeditor5-editor-inline/tests/inlineeditorui.js#L88).

I updated both of these tests to check for

```
sinon.assert.calledWithMatch( spy, 'editor-ui-deprecated-viewport-offset-config' );
```

instead of just checking for `calledOnce` for easier searching.

The only thing is that I could potentially extract it to a separate case, but it just felt natural to test this warning in one go.

> I can see the second test here, but what if _viewportOffset is never accessed for some reason? Who ensures that it is?

I have troubles finding any possible scenario when it's not accessed. Maybe after my last update to the code this became obsolete.

We do set `viewportOffset` inside `EditorUI` constructor. So every editor that has an UI will call it. If editor has no UI we don't care about setting offset data, nor warning about deprecated property usage.

File: packages/ckeditor5-core/src/editor/editorui.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (dawidurbanski) [reply packages/ckeditor5-core/src/editor/editorui.js]: Left it as a private method called `_readViewportOffsetFromConfig`

**Comment by dawidurbanski** [comment]
> * The migration guide is necessary and it doesn't need to be part of this ticket, but we need to do it. Please open a followup and ping me about it.

Done: #10401

> * It's a pity that the toolbar doesn't change its position once you update `editor.ui.viewportOffset`. I also checked if calling `editor.ui.update()` works but it doesn't. Could you make a quick check whether there's an option to update (recalculate) the position of it on demand. If so, please report a followup ticket too.

In these two tests:

```
yarn manual --files=editor-classic/manual/tickets/9672/1.js,editor-inline/manual/tickets/9672/1.js
```

We actually just set new `viewportOffset` by clicking on a button outside of the editor. This means we need to focus back in to it after it's done. We do make this step automated by adding `editor.editing.view.focus();` on button click after we are done with updating offset to new, random one.

Apparently focusing out and back in is sufficient to reposition the toolbar.

Example:

```js
const newOffset = 200;
document.documentElement.style.setProperty( '--top-offset', `${ newOffset }px` );
editor.ui.viewportOffset = { top: newOffset };
editor.editing.view.focus();
```

I added an additional, naive case to both of these tests. An another button to start updating top offset randomly in two seconds interval. This way we do keep focused in the editor while offset is being updated.

Indeed in such case editor toolbar does not get updated.

I found we just need to fire `layoutChanged` event after updating the offset.

Example:

```js
const newOffset = 200;
document.documentElement.style.setProperty( '--top-offset', `${ newOffset }px` );
editor.ui.viewportOffset = { top: newOffset };
editor.editing.view.document.fire( 'layoutChanged' );
```

In my opinion It's just a matter of proper docs to describe these cases.

**Comment by Reinmar** [comment]
> Indeed in such case editor toolbar does not get updated.

That's the case to worry about. It's clear that re-focusing resolves this issue. And it's good there's even a better workaround with `layoutChanged`. However, both of these are hacks. The straightforward solution would be if calling `editor.ui.update()` worked because that method is supposed to be called whenever the editor's UI should be updated due to some external changes. I'll write that in the issue you opened too.

**Comment by Reinmar** [comment]
FYI: The merge commit message that I went with:

> Feature: Introduced \`editor.ui.viewportOffset\` that allows modifying in runtime the viewport's offset. This value is used by the editor to e.g. position its sticky toolbar and contextual balloons. Additionally, the \`config.toolbar.viewportTopOffset\` property was moved to \`config.ui.viewportOffset\` and it now accetps an object. Closes #9672.
> 
> BREAKING CHANGE: The \`config.toolbar.viewportTopOffset\` property was moved to \`config.ui.viewportOffset\` and it now accetps an object.

**Comment by wimleers** [comment]
Looks like this also unblocked https://www.drupal.org/project/drupal/issues/3259380 per @lauriii! 🥳
